File: src/main/java/rails/game/action/BuyBonusToken.java
package rails.game.action;

import java.io.IOException;
import java.io.ObjectInputStream;

import com.google.common.base.Objects;

import net.sf.rails.game.PrivateCompany;
import net.sf.rails.game.special.SellBonusToken;
import net.sf.rails.game.special.SpecialProperty;
import net.sf.rails.game.state.Owner;
import net.sf.rails.util.RailsObjects;

/**
 * Rails 2.0: Updated equals and toString methods
 */
public class BuyBonusToken extends PossibleORAction {

    // Initial attributes
    transient private PrivateCompany privateCompany;
    private String privateCompanyName;
    transient private Owner seller = null;
    private String sellerName = null;
    transient protected SellBonusToken specialProperty = null;
    protected int specialPropertyId;

    private String name;
    private int price;
    private int value;
    private String locationString;

    public static final long serialVersionUID = 1L;

    /**
     *
     */
    public BuyBonusToken(SellBonusToken specialProperty) {

        this.specialProperty = specialProperty;
        this.specialPropertyId = specialProperty.getUniqueId();
        this.privateCompany = (PrivateCompany) specialProperty.getOriginalCompany();
        this.privateCompanyName = privateCompany.getId();
        this.seller = specialProperty.getSeller();
        if (seller != null) this.sellerName = seller.getId();
        this.name = specialProperty.getName();
        this.price = specialProperty.getPrice();
        this.value = specialProperty.getValue();
        this.locationString = specialProperty.getLocationNameString();
    }

    /**
     * @return Returns the privateCompany.
     */
    public PrivateCompany getPrivateCompany() {
        return privateCompany;
    }

    public String getPrivateCompanyName() {
        return privateCompanyName;
    }

    public Owner getSeller() {
        return seller;
    }

    public String getSellerName() {
        return sellerName;
    }

    public SellBonusToken getSpecialProperty() {
        return specialProperty;
    }

    public String getName() {
        return name;
    }

    public int getValue() {
        return value;
    }

    public String getLocationString() {
        return locationString;
    }

    public int getPrice() {
        return price;
    }

    @Override
    protected boolean equalsAs(PossibleAction pa, boolean asOption) {
        // identity always true
        if (pa == this) return true;
        //  super checks both class identity and super class attributes
        if (!super.equalsAs(pa, asOption)) return false; 

        // check asOption attributes
        BuyBonusToken action = (BuyBonusToken)pa; 
        return Objects.equal(this.privateCompany, action.privateCompany)
                && Objects.equal(this.name, action.name)
                && Objects.equal(this.price, action.price)
                && Objects.equal(this.value, action.value)
                && Objects.equal(this.locationString, action.locationString)
        ;
        // no asAction attributes to be checked
    }

    @Override
    public String toString() {
        return super.toString() + 
                RailsObjects.stringHelper(this)
                    .addToString("privateCompany", privateCompany)
                    .addToString("name", name)
                    .addToString("price", price)
                    .addToString("value", value)
                    .addToString("locationString", locationString)
                    .toString()
        ;

    }

    /** Deserialize */
    private void readObject(ObjectInputStream in) throws IOException,
            ClassNotFoundException {

        in.defaultReadObject();

        privateCompany =
                getCompanyManager().getPrivateCompany(privateCompanyName);
        if (sellerName.equalsIgnoreCase("Bank")) {
            // TODO: Assume that it is the pool, not the ipo
            seller = getRoot().getBank().getPool();
        } else if (sellerName != null) {
            seller =
                getCompanyManager().getPublicCompany(sellerName);
        }
        if (specialPropertyId > 0) {
            specialProperty =
                    (SellBonusToken) SpecialProperty.getByUniqueId(specialPropertyId);
        }
    }

}

File: dotnet/RailsLib.Net/Rails/Game/Action/BuyBonusToken.cs
using GameLib.Net.Game;
using GameLib.Net.Game.Special;
using GameLib.Net.Game.State;
using GameLib.Net.Util;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Text;

namespace GameLib.Rails.Game.Action
{
    [JsonObject(MemberSerialization.Fields)]
    public class BuyBonusToken : PossibleORAction
    {
        // Initial attributes
        /*transient*/
        [JsonIgnore]
        private PrivateCompany privateCompany;
        private string privateCompanyName;
        /*transient*/
        [JsonIgnore]
        private IOwner seller = null;
        private string sellerName = null;
        /*transient*/
        [JsonIgnore]
        protected SellBonusToken specialProperty = null;
        protected int specialPropertyId;

        private string name;
        private int price;
        private int value;
        private string locationString;

        new public const long serialVersionUID = 1L;

        /**
         *
         */
        public BuyBonusToken(SellBonusToken specialProperty)
        {

            this.specialProperty = specialProperty;
            this.specialPropertyId = specialProperty.UniqueId;
            this.privateCompany = (PrivateCompany)specialProperty.OriginalCompany;
            this.privateCompanyName = privateCompany.Id;
            this.seller = specialProperty.GetSeller();
            if (seller != null) this.sellerName = seller.Id;
            this.name = specialProperty.Name;
            this.price = specialProperty.Price;
            this.value = specialProperty.Value;
            this.locationString = specialProperty.LocationNameString;
        }

        /**
         * @return Returns the privateCompany.
         */
        public PrivateCompany PrivateCompany
        {
            get
            {
                return privateCompany;
            }
        }

        public string PrivateCompanyName
        {
            get
            {
                return privateCompanyName;
            }
        }

        public IOwner Seller
        {
            get
            {
                return seller;
            }
        }

        public string SellerName
        {
            get
            {
                return sellerName;
            }
        }

        public SellBonusToken SpecialProperty
        {
            get
            {
                return specialProperty;
            }
        }

        public string Name
        {
            get
            {
                return name;
            }
        }

        public int Value
        {
            get
            {
                return value;
            }
        }

        public string LocationString
        {
            get
            {
                return locationString;
            }
        }

        public int Price
        {
            get
            {
                return price;
            }
        }

        override protected bool EqualsAs(PossibleAction pa, bool asOption)
        {
            // identity always true
            if (pa == this) return true;
            //  super checks both class identity and super class attributes
            if (!base.EqualsAs(pa, asOption)) return false;

            // check asOption attributes
            BuyBonusToken action = (BuyBonusToken)pa;
            return privateCompany.Equals(action.privateCompany)
                    && (name == action.name)
                    && (price == action.price)
                    && (value == action.value)
                    && (locationString == action.locationString);
            // no asAction attributes to be checked
        }

        override public string ToString()
        {
            return base.ToString() +
                    RailsObjects.GetStringHelper(this)
                        .AddToString("privateCompany", privateCompany)
                        .AddToString("name", name)
                        .AddToString("price", price)
                        .AddToString("value", value)
                        .AddToString("locationString", locationString)
                        .ToString();
        }

        /** Deserialize */
        [OnDeserialized]
        new internal void OnDeserialized(StreamingContext context)
        {
            privateCompany = CompanyManager.GetPrivateCompany(privateCompanyName);
            if (sellerName.Equals("Bank", StringComparison.OrdinalIgnoreCase))
            {
                // TODO: Assume that it is the pool, not the ipo
                seller = GetRoot.Bank.Pool;
            }
            else if (sellerName != null)
            {
                seller = CompanyManager.GetPublicCompany(sellerName);
            }
            if (specialPropertyId > 0)
            {
                specialProperty = (SellBonusToken)Net.Game.Special.SpecialProperty.GetByUniqueId(GetRoot, specialPropertyId);
            }
        }
    }
}
